File: backend/src/main/java/dev/coldhands/pair/stairs/backend/usecase/UserDetailsService.java
package dev.coldhands.pair.stairs.backend.usecase;

import dev.coldhands.pair.stairs.backend.domain.User;
import dev.coldhands.pair.stairs.backend.domain.UserName;
import dev.coldhands.pair.stairs.backend.infrastructure.persistance.entity.UserEntity;
import dev.coldhands.pair.stairs.backend.infrastructure.persistance.repository.UserRepository;

import static java.util.Optional.ofNullable;

public class UserDetailsService {

    private final UserRepository userRepository;
    private final UserDisplayNameService userDisplayNameService = new UserDisplayNameService();

    public UserDetailsService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    User createOrUpdate(String oidcSub,
                        UserName userName) {
        final String displayName = userDisplayNameService.getDisplayNameFor(userName);

        final UserEntity toPersist = ofNullable(userRepository.findByOidcSub(oidcSub))
                .map(existing -> {
                    existing.setDisplayName(displayName);
                    return existing;
                })
                .orElseGet(() -> new UserEntity(oidcSub, displayName));

        final UserEntity userEntity = userRepository.saveAndFlush(toPersist);

        return new User(
                userEntity.getId(),
                userEntity.getOidcSub(),
                userEntity.getDisplayName()
        );
    }
}

File: backend/src/main/kotlin/dev/coldhands/pair/stairs/backend/usecase/UserDetailsService.kt
package dev.coldhands.pair.stairs.backend.usecase

import dev.coldhands.pair.stairs.backend.domain.OidcSub
import dev.coldhands.pair.stairs.backend.domain.user.User
import dev.coldhands.pair.stairs.backend.domain.user.UserDao
import dev.coldhands.pair.stairs.backend.domain.user.UserDetails
import dev.coldhands.pair.stairs.backend.domain.user.UserName
import dev.forkhandles.result4k.Result

class UserDetailsService(
    private val userDao: UserDao,
) {
    private val userDisplayNameService: UserDisplayNameService = UserDisplayNameService()

    fun createOrUpdate(
        oidcSub: OidcSub,
        userName: UserName,
    ): Result<User, Any> {
        val displayName = userDisplayNameService.getDisplayNameFor(userName)

        val persistedUser = userDao.findByOidcSub(oidcSub)
            ?.let { userDao.update(it.id, displayName) }
            ?: userDao.create(UserDetails(oidcSub = oidcSub, displayName = displayName))

        return persistedUser
    }

    fun getUserByOidcSub(oidcSub: OidcSub): User? =
        userDao.findByOidcSub(oidcSub)
}
